File: HTTPShortcuts/curl_command/src/main/java/ch/rmy/curlcommand/CurlConstructor.java
package ch.rmy.curlcommand;

import java.util.Map;

public class CurlConstructor {

    public static String toCurlCommandString(CurlCommand curlCommand) {
        CommandLineBuilder builder = new CommandLineBuilder("curl");

        builder.argument(curlCommand.getUrl());

        if (!"GET".equals(curlCommand.getMethod())) {
            builder.option("-X", curlCommand.getMethod());
        }

        if (curlCommand.getTimeout() != 0) {
            builder.option("-m", curlCommand.getTimeout());
        }

        if (!curlCommand.getUsername().isEmpty()) {
            builder.option("-u", curlCommand.getUsername() + (!curlCommand.getPassword().isEmpty() ? ":" + curlCommand.getPassword() : ""));
        }

        for (Map.Entry<String, String> header : curlCommand.getHeaders().entrySet()) {
            builder.option("-H", header.getKey() + ": " + header.getValue());
        }

        if (!curlCommand.getData().isEmpty()) {
            builder.option("-d", curlCommand.getData());
        }

        return builder.build();
    }

}

File: HTTPShortcuts/curl_command/src/main/kotlin/ch/rmy/curlcommand/CurlConstructor.kt
package ch.rmy.curlcommand

object CurlConstructor {

    fun toCurlCommandString(curlCommand: CurlCommand): String {
        val builder = CommandLineBuilder("curl")

        builder.argument(curlCommand.url)

        if (curlCommand.method != "GET") {
            builder.option("-X", curlCommand.method)
        }

        if (curlCommand.timeout != 0) {
            builder.option("-m", curlCommand.timeout)
        }

        if (curlCommand.username.isNotEmpty()) {
            if (curlCommand.isDigestAuth) {
                builder.option("--digest")
            }
            builder.option("-u", curlCommand.username + if (curlCommand.password.isNotEmpty()) ":" + curlCommand.password else "")
        }

        if (curlCommand.proxyHost.isNotEmpty()) {
            builder.option("-x", "${curlCommand.proxyHost}:${curlCommand.proxyPort}")
        }

        for ((key, value) in curlCommand.headers) {
            builder.option("-H", "$key: $value")
        }

        curlCommand.data.forEach { data ->
            if (curlCommand.isFormData) {
                builder.option("-F", data)
            } else {
                builder.option("-d", data)
            }
        }

        if (curlCommand.usesBinaryData) {
            builder.option("--data-binary", "@file")
        }

        if (curlCommand.insecure) {
            builder.option("--insecure")
        }

        if (curlCommand.silent) {
            builder.option("--silent")
        }

        if (curlCommand.ipVersion4) {
            builder.option("--ipv4")
        }

        if (curlCommand.ipVersion6) {
            builder.option("--ipv6")
        }

        return builder.build()
    }
}
